awx/api/views/adhoc_workflow.py

```diff
@@ -0,0 +1,192 @@
+
+        if len(inventory_ids) > 0:
+            inventory_ids = set(inventory_ids)
+            if not request.user.is_superuser:
+                accessible_inventory_ids = inventory_ids.intersection({item[0] for item in Inventory.accessible_pk_qs(request.user, 'read_role').all()})
+            else:
+                accessible_inventory_ids = inventory_ids
+            invs = request.user.get_queryset(Inventory).filter(id__in=accessible_inventory_ids)
+            inv_map = {inv.id: inv for inv in invs}
+            if inventory_ids != set(inv_map.keys()):
+                node_errors.append(f"Unable to locate or lack permissions to inventory objects: {set(inventory_ids) - set(inv_map.keys())}")
+        if len(credential_ids) > 0:
+            credential_ids = set(credential_ids)
+            if not request.user.is_superuser:
+                accessible_credential_ids = credential_ids.intersection({item[0] for item in Credential.accessible_pk_qs(request.user, 'read_role').all()})
+            else:
+                accessible_credential_ids = credential_ids
+            creds = request.user.get_queryset(Credential).filter(id__in=accessible_credential_ids)
+            cred_map = {cred.id: cred for cred in creds}
+            if credential_ids != set(cred_map.keys()):
+                node_errors.append(f"Unable to locate or lack permissions to credentials: {set(credential_ids) - set(cred_map.keys())}")
+
+        if len(node_errors) > 0:
+            return Response(' '.join(node_errors), status=status.HTTP_400_BAD_REQUEST)
+
+        # Now that we have validated the workflow job node input and acess to underlying resources,
+        # we can create the workflow job and the workflow nodes
+        name = request.data['name'] if 'name' in request.data else 'AdHocWorkflow'
+        wfj = WorkflowJob.objects.create(name=name)
+        for kwargs in workflow_node_data:
+            kwargs['unified_job_template'] = ujt_map[kwargs['unified_job_template']]
+            if 'inventory' in kwargs:
+                kwargs['inventory'] = inv_map[kwargs['inventory']]
+            if 'credentials' in kwargs:
+                # Have to set m2m fields like credentials after node is created
+                creds = [cred_map[cred_id] for cred_id in kwargs['credentials']]
+                node_m2m_objects[kwargs['identifier']]['credentials'] = creds
+                kwargs.pop('credentials')
+            wf_node = WorkflowJobNode(workflow_job=wfj, created=wfj.created, modified=wfj.modified, **kwargs)
+            workflow_nodes.append(wf_node)
+
+        try:
+            workflow_nodes = WorkflowJobNode.objects.bulk_create(workflow_nodes)
+            for node in workflow_nodes:
+                if 'credentials' in node_m2m_objects[node.identifier]:
+                    node.credentials.set(node_m2m_objects[node.identifier]['credentials'])
```

Comment:
@benthomasson has mentioned we could do a bulk insert to a many to many table to save this N number of inserts that we have to do for each node that has

https://docs.djangoproject.com/en/4.1/ref/models/options/#db-table